unittests/SILOptimizer/PartitionUtilsTest.cpp

```diff
@@ -100,6 +125,235 @@ TEST(PartitionUtilsTest, TestMergeAndJoin) {
+  p2.apply(PartitionOp::Assign(Element(25), Element(25)));
+  p2.apply(PartitionOp::Assign(Element(26), Element(31)));
+  p2.apply(PartitionOp::Assign(Element(27), Element(16)));
+  p2.apply(PartitionOp::Assign(Element(28), Element(35)));
+  p2.apply(PartitionOp::Assign(Element(29), Element(40)));
+  p2.apply(PartitionOp::Assign(Element(30), Element(33)));
+  p2.apply(PartitionOp::Assign(Element(31), Element(34)));
+  p2.apply(PartitionOp::Assign(Element(32), Element(22)));
+  p2.apply(PartitionOp::Assign(Element(33), Element(42)));
+  p2.apply(PartitionOp::Assign(Element(34), Element(37)));
+  p2.apply(PartitionOp::Assign(Element(35), Element(34)));
+  p2.apply(PartitionOp::Assign(Element(36), Element(18)));
+  p2.apply(PartitionOp::Assign(Element(37), Element(32)));
+  p2.apply(PartitionOp::Assign(Element(38), Element(22)));
+  p2.apply(PartitionOp::Assign(Element(39), Element(44)));
+  p2.apply(PartitionOp::Assign(Element(40), Element(20)));
+  p2.apply(PartitionOp::Assign(Element(41), Element(37)));
+  p2.apply(PartitionOp::Assign(Element(43), Element(29)));
+  p2.apply(PartitionOp::Assign(Element(44), Element(25)));
+
+  auto result = Partition::join(p1, p2);
+  PartitionTester tester(result);
+  ASSERT_EQ(tester.getRegion(0), 0);
+  ASSERT_EQ(tester.getRegion(1), 1);
+  ASSERT_EQ(tester.getRegion(2), 0);
+  ASSERT_EQ(tester.getRegion(3), 3);
+  ASSERT_EQ(tester.getRegion(4), 4);
+  ASSERT_EQ(tester.getRegion(5), 5);
+  ASSERT_EQ(tester.getRegion(6), 6);
+  ASSERT_EQ(tester.getRegion(7), 3);
+  ASSERT_EQ(tester.getRegion(8), 3);
+  ASSERT_EQ(tester.getRegion(9), 3);
+  ASSERT_EQ(tester.getRegion(10), 3);
+  ASSERT_EQ(tester.getRegion(11), 11);
+  ASSERT_EQ(tester.getRegion(12), 0);
+  ASSERT_EQ(tester.getRegion(13), 13);
+  ASSERT_EQ(tester.getRegion(14), 1);
+  ASSERT_EQ(tester.getRegion(15), 13);
+  ASSERT_EQ(tester.getRegion(16), 0);
+  ASSERT_EQ(tester.getRegion(17), 3);
+  ASSERT_EQ(tester.getRegion(18), 13);
+  ASSERT_EQ(tester.getRegion(19), 13);
+  ASSERT_EQ(tester.getRegion(20), 13);
+  ASSERT_EQ(tester.getRegion(21), 3);
+  ASSERT_EQ(tester.getRegion(22), 13);
+  ASSERT_EQ(tester.getRegion(23), 3);
+  ASSERT_EQ(tester.getRegion(24), 1);
+  ASSERT_EQ(tester.getRegion(25), 3);
+  ASSERT_EQ(tester.getRegion(26), 13);
+  ASSERT_EQ(tester.getRegion(27), 0);
+  ASSERT_EQ(tester.getRegion(28), 3);
+  ASSERT_EQ(tester.getRegion(29), 13);
+  ASSERT_EQ(tester.getRegion(30), 1);
+  ASSERT_EQ(tester.getRegion(31), 0);
+  ASSERT_EQ(tester.getRegion(32), 13);
+  ASSERT_EQ(tester.getRegion(33), 33);
+  ASSERT_EQ(tester.getRegion(34), 34);
+  ASSERT_EQ(tester.getRegion(35), 34);
+  ASSERT_EQ(tester.getRegion(36), 13);
+  ASSERT_EQ(tester.getRegion(37), 13);
+  ASSERT_EQ(tester.getRegion(38), 13);
+  ASSERT_EQ(tester.getRegion(39), 39);
+  ASSERT_EQ(tester.getRegion(40), 13);
+  ASSERT_EQ(tester.getRegion(41), 13);
+  ASSERT_EQ(tester.getRegion(42), 33);
+  ASSERT_EQ(tester.getRegion(43), 13);
+  ASSERT_EQ(tester.getRegion(44), 3);
```

Comment:
A suggestion for future testing - random number of elements assigned randomly, we could keep track of all of these assignments in a map and double-check that join is correct via double-checking that map in a loop.

Reply:
@xedin I did that here. I just generated it offline.